File: 0446_arithmetic_slices_ii_subsequence/solution.rb
# LeetCode 0446 - Arithmetic Slices II - Subsequence
# https://leetcode.com/problems/arithmetic-slices-ii-subsequence/

class Solution
  def number_of_arithmetic_slices(nums)
    total = 0
    differences = Array.new(nums.length) { Hash.new(0) }

    nums.each_with_index do |value, index|
      (0...index).each do |previous|
        diff = value - nums[previous]
        total += differences[previous][diff]
        differences[index][diff] += differences[previous][diff] + 1
      end
    end

    total
  end

  alias_method :numberOfArithmeticSlices, :number_of_arithmetic_slices
end

File: 0446_arithmetic_slices_ii_subsequence/solution.php
// LeetCode 0446 - Arithmetic Slices II - Subsequence
// https://leetcode.com/problems/arithmetic-slices-ii-subsequence/

class Solution {
    /**
     * @param int[] $nums
     * @return int
     */
    function numberOfArithmeticSlices($nums) {
        return $this->number_of_arithmetic_slices($nums);
    }

    /**
     * @param int[] $nums
     * @return int
     */
    function number_of_arithmetic_slices($nums) {
        $total = 0;
        $differences = array_fill(0, count($nums), []);

        foreach ($nums as $index => $value) {
            for ($previous = 0; $previous < $index; $previous++) {
                $diff = $value - $nums[$previous];
                $total += $differences[$previous][$diff] ?? 0;
                $differences[$index][$diff] = ($differences[$index][$diff] ?? 0) + ($differences[$previous][$diff] ?? 0) + 1;
            }
        }

        return $total;
    }
}
